File: src/AppPaths.cpp
#include "AppPaths.h"

#include <QDir>
#include <QFile>
#include <QFileDevice>
#include <QStandardPaths>

namespace {
QString fallbackAppDataPath()
{
    const QString home = QDir::homePath();
    return home + QStringLiteral("/.local/share/MatrixMediaShareClientQt");
}
}

AppPaths::AppPaths()
{
    QString base = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
    if (base.isEmpty()) {
        base = fallbackAppDataPath();
    }

    rootPath_ = base + QStringLiteral("/MatrixMediaShareClientQt");
    appSupportPath_ = rootPath_;
    databasePath_ = rootPath_ + QStringLiteral("/app.sqlite");
    matrixDataPath_ = rootPath_ + QStringLiteral("/matrix-sdk/data");
    matrixCachePath_ = rootPath_ + QStringLiteral("/matrix-sdk/cache");
    tempDownloadsPath_ = rootPath_ + QStringLiteral("/tmp-downloads");
    libraryPath_ = rootPath_ + QStringLiteral("/shared-files");
    manualDownloadsPath_ = rootPath_ + QStringLiteral("/downloads");
    landingPagesPath_ = rootPath_ + QStringLiteral("/gateway-pages");
    kuboPath_ = rootPath_ + QStringLiteral("/kubo");
    kuboRepoPath_ = rootPath_ + QStringLiteral("/kubo/repo");
    secretStorePath_ = rootPath_ + QStringLiteral("/secrets.json");

    ensureDirectory(rootPath_);
    ensureDirectory(appSupportPath_);
    ensureDirectory(matrixDataPath_);
    ensureDirectory(matrixCachePath_);
    ensureDirectory(tempDownloadsPath_);
    ensureDirectory(libraryPath_);
    ensureDirectory(manualDownloadsPath_);
    ensureDirectory(landingPagesPath_);
    ensureDirectory(kuboPath_);
    ensureDirectory(kuboRepoPath_);
}

QString AppPaths::rootPath() const
{
    return rootPath_;
}

QString AppPaths::appSupportPath() const
{
    return appSupportPath_;
}

QString AppPaths::databasePath() const
{
    return databasePath_;
}

QString AppPaths::matrixDataPath() const
{
    return matrixDataPath_;
}

QString AppPaths::matrixCachePath() const
{
    return matrixCachePath_;
}

QString AppPaths::tempDownloadsPath() const
{
    return tempDownloadsPath_;
}

QString AppPaths::libraryPath() const
{
    return libraryPath_;
}

QString AppPaths::manualDownloadsPath() const
{
    return manualDownloadsPath_;
}

QString AppPaths::landingPagesPath() const
{
    return landingPagesPath_;
}

QString AppPaths::kuboPath() const
{
    return kuboPath_;
}

QString AppPaths::kuboRepoPath() const
{
    return kuboRepoPath_;
}

QString AppPaths::secretStorePath() const
{
    return secretStorePath_;
}

void AppPaths::ensureDirectory(const QString &path) const
{
    QDir directory;
    directory.mkpath(path);
    lockDownPermissions(path);
}

void AppPaths::lockDownPermissions(const QString &path) const
{
#ifdef Q_OS_UNIX
    QFile::setPermissions(path, QFileDevice::ReadOwner | QFileDevice::WriteOwner | QFileDevice::ExeOwner);
#else
    Q_UNUSED(path);
#endif
}

File: backend/src/app_paths.rs
use std::{
    fs,
    path::{Path, PathBuf},
};

use anyhow::{Context, Result};

#[derive(Clone, Debug)]
pub struct AppPaths {
    pub root_path: PathBuf,
    pub database_path: PathBuf,
    pub matrix_data_path: PathBuf,
    pub matrix_cache_path: PathBuf,
    pub temp_downloads_path: PathBuf,
    pub thumbnail_cache_path: PathBuf,
    pub library_path: PathBuf,
    pub manual_downloads_path: PathBuf,
    pub landing_pages_path: PathBuf,
    pub kubo_path: PathBuf,
    pub kubo_repo_path: PathBuf,
    pub secret_store_path: PathBuf,
}

impl AppPaths {
    pub fn from_root(root_path: PathBuf) -> Self {
        Self {
            database_path: root_path.join("app.sqlite"),
            matrix_data_path: root_path.join("matrix-sdk").join("data"),
            matrix_cache_path: root_path.join("matrix-sdk").join("cache"),
            temp_downloads_path: root_path.join("tmp-downloads"),
            thumbnail_cache_path: root_path.join("thumbnail-cache"),
            library_path: root_path.join("shared-files"),
            manual_downloads_path: root_path.join("downloads"),
            landing_pages_path: root_path.join("gateway-pages"),
            kubo_path: root_path.join("kubo"),
            kubo_repo_path: root_path.join("kubo").join("repo"),
            secret_store_path: root_path.join("secrets.json"),
            root_path,
        }
    }

    pub fn ensure_directories(&self) -> Result<()> {
        self.ensure_directory(&self.root_path)?;
        self.ensure_directory(&self.matrix_data_path)?;
        self.ensure_directory(&self.matrix_cache_path)?;
        self.ensure_directory(&self.temp_downloads_path)?;
        self.ensure_directory(&self.thumbnail_cache_path)?;
        self.ensure_directory(&self.library_path)?;
        self.ensure_directory(&self.manual_downloads_path)?;
        self.ensure_directory(&self.landing_pages_path)?;
        self.ensure_directory(&self.kubo_path)?;
        self.ensure_directory(&self.kubo_repo_path)?;
        Ok(())
    }

    fn ensure_directory(&self, path: &Path) -> Result<()> {
        fs::create_dir_all(path)
            .with_context(|| format!("Failed to create directory {}", path.display()))?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;

            let permissions = fs::Permissions::from_mode(0o700);
            fs::set_permissions(path, permissions)
                .with_context(|| format!("Failed to secure directory {}", path.display()))?;
        }
        Ok(())
    }
}
